### PR TITLE
docs(python): fix and clarify docs for `Expr.map_elements`

### PR DESCRIPTION
**Pull request by MarcoGorelli**
I'm suggesting to clarify the `Expr.apply` docs.

First, because `Expr.apply` is very simple and predictable, but I think the current docs make it look a bit magical and that its behaviour depends on the context.
I think that by showing the pre-aggregation that happens in groupby, then Expr.apply can be simply described as acting per-element

But second, because I'm not sure the current docs are quite correct. They say that in a groupby context, it

> Expects `f` to be of type Callable[[Series], Series].

Is this right? E.g. if we have
```python
In [11]: df = pl.DataFrame({"a": [1, 1, 2], 'b': [4,5,6]})

In [12]: df.groupby('a').agg('b')
Out[12]:
shape: (2, 2)
┌─────┬───────────┐
│ a   ┆ b         │
│ --- ┆ ---       │
│ i64 ┆ list[i64] │
╞═════╪═══════════╡
│ 2   ┆ [6]       │
│ 1   ┆ [4, 5]    │
└─────┴───────────┘
```
and then do
```python
In [13]: df.groupby('a').agg(pl.col('b').apply(lambda x: x.sum()))
Out[13]:
shape: (2, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 1   ┆ 9   │
│ 2   ┆ 6   │
└─────┴─────┘
```
then the function first took in the series
```python
shape: (2,)
Series: '' [i64]
[
        4
        5
]
```
and returned `9`, and then it took the series
```python
shape: (1,)
Series: '' [i64]
[
        6
]
```
and returned `6`.

So it's true that it accepts a `Series`, but it doesn't have to return one.

### CODE REVIEW

File: py-polars/polars/expr/expr.py [hunk 1]
```diff
@@ -3783,16 +3778,37 @@ def apply(
...
+                3
+                1
+        ]
+
+        Therefore, from the user's point-of-view, the function is applied by group:
 
-        >>> df.lazy().groupby("b", maintain_order=True).agg(
-        ...     pl.col("a").apply(lambda x: x.sum())
+        >>> df.lazy().groupby("b").agg(pl.col("a").apply(lambda x: x.sum())).sort(
+        ...     "b"
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/expr.py:3811]: minor drive-by - I think it would be better to not teach `maintain_order=True`, as that's more expensive. the example can be made deterministic with `sort`, and then at least it should be clearer that that part is optional, and people will default to the more performant `maintain_order=False`

Comment (stinodego) [reply py-polars/polars/expr/expr.py:3811]: I agree we shouldn't be including `maintain_order` in every groupby example. However, adding `sort` everywhere adds additional clutter that I'd rather not have in the examples.

I'd vote to simply add a `# doctest: +IGNORE_RESULT` to these examples.

File: py-polars/polars/expr/expr.py [hunk 2]
```diff
@@ -3783,16 +3778,37 @@ def apply(
...
+        │ b   ┆ [2]       │
+        │ c   ┆ [3, 1]    │
+        └─────┴───────────┘
+        >>> df.lazy().groupby("b").agg(pl.col("a")).sort("b").collect()["a"][-1]
+        shape: (2,)
+        Series: '' [i64]
+        [
+                3
+                1
+        ]
```
Comment (stinodego) [inline py-polars/polars/expr/expr.py:3806]: I think this can be omitted.

File: py-polars/polars/expr/expr.py [hunk 3]
```diff
@@ -3702,20 +3702,15 @@ def apply(
         strategy: ApplyStrategy = "thread_local",
     ) -> Self:
         """
-        Apply a custom/user-defined function (UDF) in a GroupBy or Projection context.
+        Apply a custom/user-defined function (UDF) to each element of a Series.
```
Comment (stinodego) [inline py-polars/polars/expr/expr.py:3705]: When I was newer to Polars, it was confusing to me when expressions were talking about operating on Series. Now I now that expressions operate on Series in the background, but maybe we can avoid potential confusion by simply referring to 'columns' here?

```suggestion
        Apply a custom/user-defined function (UDF) to each element of a column.
```

**Review by stinodego — COMMENTED** [review]
Some minor comments, but I think this is a nice improvement!

**Comment by stinodego** [comment]
> Expects f to be of type Callable[[Series], Series].

As you've found, this is not correct. You could amend this and distinguish between use in a regular vs a groupby context.

File: py-polars/polars/expr/expr.py [hunk 4]
```diff
@@ -3733,14 +3737,14 @@ def apply(
...
             - 'thread_local': run the python function on a single thread.
             - 'threading': run the python function on separate threads. Use with
-                        care as this can slow performance. This might only speed up
-                        your code if the amount of work per element is significant
-                        and the python function releases the GIL (e.g. via calling
-                        a c function)
+              care as this can slow performance. This might only speed up
+              your code if the amount of work per element is significant
+              and the python function releases the GIL (e.g. via calling
+              a c function)
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/expr.py:3747]: this isn't showing up correctly in the[ released docs](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.apply.html)

here, they look like this:

image

**Comment by MarcoGorelli** [comment]
thanks for your review! have addressed the feedback (+ merged main and updated `groupby` -> `group_by`)

**Comment by ritchie46** [comment]
@MarcoGorelli I think this will collide with #10678. Care to re-open these changes once #10678 is merged? :)

**Comment by MarcoGorelli** [comment]
Sounds good!

**Review by stinodego — APPROVED** [review]
Thanks!
